src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -0,0 +1,1011 @@
+
+        private readonly SemaphoreSlim _writerLock;
+
+        private int _nextStream;
+        private bool _expectingSettingsAck;
+
+        private const int MaxStreamId = int.MaxValue;
+
+        private static readonly byte[] s_http2ConnectionPreface = Encoding.ASCII.GetBytes("PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n");
+
+        private const int InitialBufferSize = 4096;
+
+        public Http2Connection(SslStream stream)
+        {
+            _stream = stream;
+            _incomingBuffer = new ArrayBuffer(InitialBufferSize);
+            _outgoingBuffer = new ArrayBuffer(InitialBufferSize);
+
+            _hpackDecoder = new HPackDecoder();
+
+            _httpStreams = new ConcurrentDictionary<int, Http2Stream>();
+
+            _writerLock = new SemaphoreSlim(1);
+
+            _nextStream = 1;
+        }
+
+        public async Task SetupAsync()
+        {
+            // Send connection preface
+            _outgoingBuffer.EnsureWriteSpace(s_http2ConnectionPreface.Length);
+            s_http2ConnectionPreface.AsSpan().CopyTo(_outgoingBuffer.WriteBytes);
+            _outgoingBuffer.Commit(s_http2ConnectionPreface.Length);
+
+            // Send empty settings frame
+            WriteFrameHeader(new FrameHeader(0, FrameType.Settings, FrameFlags.None, 0));
+
+            // TODO: We should disable PUSH_PROMISE here.
+            // TODO: We should send a connection-level WINDOW_UPDATE to allow
+            // a large amount of data to be received on the connection.  
+            // We don't care that much about connection-level flow control, we'll manage it per-stream.
+
+            await _stream.WriteAsync(_outgoingBuffer.ReadMemory).ConfigureAwait(false);
+            _outgoingBuffer.Consume(_outgoingBuffer.ReadMemory.Length);
+
+            _expectingSettingsAck = true;
+
+            // Receive the initial SETTINGS frame from the peer.
+            FrameHeader frameHeader = await ReadFrameAsync().ConfigureAwait(false);
+            if (frameHeader.Type != FrameType.Settings || frameHeader.AckFlag)
+            {
+                throw new Http2ProtocolException(Http2ProtocolErrorCode.ProtocolError);
+            }
+
+            // Process the SETTINGS frame.  This will send an ACK.
+            await ProcessSettingsFrame(frameHeader).ConfigureAwait(false);
+
+            ProcessIncomingFrames();
+        }
+
+        private int GetNewStreamId()
+        {
+            int newStream = _nextStream;
+            while (true)
+            {
+                if (newStream == MaxStreamId)
+                {
+                    // No more stream Ids available
+                    return -1;
+                }
+
+                int original = Interlocked.CompareExchange(ref _nextStream, newStream + 2, newStream);
+                if (original == newStream)
+                {
+                    return newStream;
+                }
+
+                newStream = original;
+            }
```

Comment:
This implementation suggests that after ~1 billion requests on the same connection, we will no longer be able to make HTTP requests to that server?  I realize that's a large number, but it's also small enough to be feasible for a high-throughput server that's up for a long time.

Reply:
See above. The intended behavior is to tear down the existing connection and establish a new one.

Reply:
> The intended behavior is to tear down the existing connection and establish a new one.

Ok.  That's something I think we're going to need to implement before shipping. An issue / TODO or whatever would be good.

Reply:
Agreed. There's a TODO at the call site in SendAsync.